feat: add ability to define interceptors

StephaneRob opened the pull request:
This PR add ability to define interceptors for a given mailer. An interceptor can modify or intercept email before send.

germsvel commented:
@StephaneRob thanks so much for opening this PR. I'll review it as soon as I can. 

In the meantime, I was wondering if you could provide some more information as to the possible use cases for interceptors? I know I've used similar things in the past to prevent staging emails from going out, so I can see how this would be very useful. 

But I'd love to have more information so we can make sure to document all of that. Thanks!

StephaneRob commented:
Hi @germsvel,
As you said, interceptors can either be used to prevent emails from going out or to modify some email data such as subject in a single place.

use cases for interceptions:
- block outbound emails for a given environment (staging, dev,...)
- block email in case the recipient is blocked (bounces, complaints, ...)

use cases for modifications:
- rewrite subject for a given environment (for example I like prefix subject w/ environment ex: `[staging] Welcome`. the subject is `Welcome` and the mailer take care of this part, and the interceptor will prefix when needed.
- rewrite email to make sure no email is sent to real email adress in a dev environment.

Certainly others use cases :)

germsvel commented on test/lib/bamboo/mailer_test.exs:
```diff
@@ -447,6 +453,26 @@ defmodule Bamboo.MailerTest do
+      Application.put_env(:bamboo, __MODULE__.MailerWithInterceptors, @mailer_config)
+    end
+
+    test "must apply interceptor and send email if not intercepted" do
```
Thank you for adding these tests for `deliver_now` 🎉  

I like that we're testing two conditions: when the interceptor modifies the email (this test) and when the interceptor blocks the email (the next one). 👍

germsvel reviewed (commented):
Thank you so much for introducing this work.  I'm really excited for this. I think this will be very useful! 🥳

I left several comments. If any of them are unclear, please let me know. I'd be happy to clarify them or help as much as I can. 

I think the biggest change needed is for us to handle all `deliver_x` functions. The PR currently handles `deliver_now`. 

Thanks again, and please reach out if you have comments or questions.

StephaneRob commented:
@germsvel thank you for your review :)
Happy that this feature can be useful, i'm working on it and let you know as soon as a new version is available

StephaneRob commented on mix.exs:
```diff
@@ -67,7 +67,8 @@ defmodule Bamboo.Mixfile do
       {:hackney, ">= 1.15.2"},
-      {:jason, "~> 1.0", optional: true}
+      {:jason, "~> 1.0", optional: true},
+      {:mix_test_watch, "~> 1.0", only: :dev, runtime: false}
```
👌

StephaneRob commented on README.md:
````diff
@@ -260,6 +260,38 @@ struct directly to Bamboo anywhere it expects an address. See the
+
+```elixir
+# some/path/within/your/app/black_list_interceptor.ex
+defmodule MyApp.BlackListInterceptor do
````
Of course!
(Maybe  moving the `master` branch to `main` could be a good idea to go further)

StephaneRob commented on lib/bamboo/mailer.ex:
```diff
@@ -67,29 +67,36 @@ defmodule Bamboo.Mailer do
               | {:ok, Bamboo.Email.t(), any}
               | {:error, Exception.t() | String.t()}
+
+      interceptors = Keyword.get(opts, :interceptors, [])
```
Yes good idea!

StephaneRob commented on test/lib/bamboo/mailer_test.exs:
```diff
@@ -447,6 +453,26 @@ defmodule Bamboo.MailerTest do
+    end
+
+    test "must apply interceptor and block email if intercepted" do
+      email = new_email(to: "bar@foo.com")
```
👌

StephaneRob commented on lib/bamboo/interceptor.ex:
```diff
@@ -0,0 +1,10 @@
+  defmacro __using__(_) do
+    quote do
+      @behaviour Bamboo.Interceptor
+      def call(:intercepted), do: :intercepted
```
The call is here only to handle `intercepted` email in next interceptors only by using `Bamboo.Interceptor`. It's also possible to let the user handle it by himself. What do you think?

(it was the only purpose of `__using__` macro, below)

StephaneRob commented on lib/bamboo/interceptor.ex:
```diff
@@ -0,0 +1,10 @@
+defmodule Bamboo.Interceptor do
+  @callback call(email :: Bamboo.Email.t()) :: Bamboo.Email.t() | :intercepted
+
+  defmacro __using__(_) do
```
If we remove the fallback `call` (above), we can remove it.

StephaneRob commented on lib/bamboo/mailer.ex:
```diff
@@ -194,6 +201,8 @@ defmodule Bamboo.Mailer do
   end
 
   @doc false
+  def deliver_now(_, :intercepted, _, _), do: nil
```
Yes my mistake sorry!

What about applying interceptors like this in `deliver_now` and `deliver_later` ?

```elixir
  def deliver_later(adapter, email, config) do
    with %Bamboo.Email{} = email <- apply_interceptors(email, config),
         {:ok, email} <- validate_and_normalize(email, adapter) do
         #...
 ```

germsvel replied:
👍  I think those would be the best functions to handle that.

StephaneRob commented:
Hi @germsvel, I reworked the feature. I also moved `apply_interceptors` after `validate_and_normalize`. It allows to work with a single format of `from`, `to`, `cc` and `bcc`.

Let me know if it's more like you imagined it

davidlibrera commented:
Nice work. :D
I was looking exactly for this feature and I've created a branch with a similiar feature.

@StephaneRob  what do you think about adding some headers with original addresses? That way it's possible to see which addresses should have received emails without interception.
I've imaged something like

```
%{
  "X-Real-To" => list_of_real_to_recipients,
  "X-Real-Cc" => list_of_real_cc_recipients,
  "X-Real-Bcc" => list_of_real_bcc_recipients
}
```

StephaneRob commented:
Thx @davidlibrera :)
Interceptors is something generic, Bamboo doesn't do anything by default (logic will be on application side). In your use case you could create a `RecipientReplacerInterceptor` which replace recipients and add headers w/ original recipients.

davidlibrera commented:
Ok, I'm understand what's happening. This PR was notified to me as the same feature I was working by @maartenvanvliet but it seems that provides a different feature.

StephaneRob commented:
@davidlibrera At first sight, your use case fits with interceptors feature.

davidlibrera commented:
Ok, my fault.
I didn't notice that the `DenyListInterceptor` is a test module. Ok, now I figured out how it works and I imagine that in order to have what I was trying to achieve I simply need to create an Interceptor that replaces recipients. Am I right?

StephaneRob commented:
Yes exactly :)

davidlibrera commented:
It sounds good! Many thanks!

germsvel reviewed (commented):
Thank you for all the improvements! 🥳 

I left a few more comments, because I think we might not be handling two cases that we need to handle.

germsvel commented:
As I've read through the code, one thing that stood out to me is that we're using the term _intercept_ for two different (but related) concepts. And that can make it confusing. 

One the one hand, we can define an _interceptor_ module that _modifies_ the email on its way out. A good example of that is the `EnvInterceptor` you've defined in the tests &mdash; it adds the environment to the subject. 

On the other hand, we can define _interceptor_ modules that _block_ an email from going out. That's when we use the `Email.intercept/1` function to mark the `%Email{intercepted: true}`. 

So, I think the part that could cause confusion is that you can have an email that has been "intercepted" in the sense that it's gone through an interceptor but that hasn't been blocked &mdash; and thus has `intercepted: false`. 

So, here's my question. What do you think about changing the term we use for one of those? I think the term _interceptor_ implies blocking or obstructing, so it makes sense that an email is intercepted. So perhaps we can find a better name for the modules? 

But please feel free to disagree with this. It's possible that the term _interceptor_ is good enough for both the modules and the boolean field. I'd love to hear your thoughts and see if we can find other terms. If we can't find other terms that better express that, I think I'm okay with using _intercept_ for both.

StephaneRob commented on test/lib/bamboo/mailer_test.exs:
```diff
@@ -447,6 +451,27 @@ defmodule Bamboo.MailerTest do
+      end
+
+      @tag interceptors: [Bamboo.DenyListInterceptor, Bamboo.EnvInterceptor]
+      test "#{method}/1 &must apply interceptor and block email if intercepted" do
```
Actually this is a typo, it is `must`

StephaneRob commented on lib/bamboo/email.ex:
```diff
@@ -263,4 +265,8 @@ defmodule Bamboo.Email do
   end
+
+  def intercept(email) do
+    %{email | intercepted: true}
```
👍

StephaneRob commented on lib/bamboo/email.ex:
```diff
@@ -83,7 +83,8 @@ defmodule Bamboo.Email do
           assigns: %{atom => any},
-          private: %{atom => any}
+          private: %{atom => any},
+          intercepted: boolean()
```
Maybe we loose the ability to have a default value on `intercepted`. If we move it to private, we can also add a method to check if the email is intercepted by looking for `:intercepted` in private ?

germsvel replied:
> Maybe we loose the ability to have a default value on intercepted.

Yeah, that's a good point. I like having that default. Let's keep the key at the top level then. I think it'll be fine.

StephaneRob commented on lib/bamboo/mailer.ex:
```diff
@@ -195,7 +196,8 @@ defmodule Bamboo.Mailer do
   def deliver_now(adapter, email, config, opts) do
-    with {:ok, email} <- validate_and_normalize(email, adapter) do
+    with {:ok, email} <- validate_and_normalize(email, adapter),
+         %Bamboo.Email{intercepted: false} = email <- apply_interceptors(email, config) do
```
I agree with you, if an email is intercepted it's not really an error but more an intented behaviour. I would go for `{:ok, email}` or `{:ok, email, "intercepted"}`.

germsvel replied:
Excellent. 👍  Let's make sure we return `{:ok, email`} when an email has been intercepted then.

StephaneRob commented:
I understand what you mean. The term interceptor comes originally from Action Mailer (rails). They use Interceptor for the class but there is not `intercept` notion (only prevent delivery via `mail.perform_deliveries = false`.

Maybe we can change this term to something closer than a `Hook` ? 

```elixir
config :my_app, MyApp.Mailer,
  adapter: MyAdapter,
  before_send: [EnvHook, DenyListHook]
  
# with 
defmodule Bamboo.Hook do
  @callback call(email :: Bamboo.Email.t()) :: Bamboo.Email.t()
end
```

`before_send` would receive a list of adopted `Hook` behaviours.

germsvel commented:
@StephaneRob thanks for mentioning that. I took a look at some of the interceptor libraries in Ruby, and I also tried to look for similar language used by other projects (for example, [intercepting outgoing events in Phx Channels](https://hexdocs.pm/phoenix/Phoenix.Channel.html#module-intercepting-outgoing-events)). It seems that the concept of _intercept_ is used not only for blocking. 

So, now I think we should keep the `interceptors` config key and the module names as `Interceptor`. But what do you think about changing the `%Email{intercepted: true}` key to something like `blocked`? (e.g.`%Email{blocked: true})`. That way there's a distinction between emails being intercepted (and modified) from those that are blocked. 

I think that also lines up with the language you used in the docs for the README: 

> It's possible to configure per Mailer interceptors. Interceptors allow
to modify / intercept (block) email on the fly.

I think that's a great description of interceptors: "Interceptors allow you to modify / block email on the fly. 

What do you think?

StephaneRob commented:
@germsvel good idea! I update the pull request and let you know when ready for review.

StephaneRob commented:
@germsvel I updated the PR.

germsvel commented on test/lib/bamboo/mailer_test.exs:
```diff
@@ -447,6 +451,76 @@ defmodule Bamboo.MailerTest do
+      assert %Bamboo.Email{blocked: true} = Mailer.deliver_later!(email)
+      refute_receive {:deliver, %Bamboo.Email{to: [{nil, "blocked@blocked.com"}]}, _config}
+    end
+  end
```
I like the change to these tests. They are far easier to read 🎉

germsvel reviewed (approved):
This looks great! Thank you so much for all the work and back and forth @StephaneRob! 🥳 

Very excited for this feature.

StephaneRob commented:
@germsvel thank you for your help!
